runner/scheduler.py

```diff
@@ -166,34 +172,39 @@ def send_message(self, test_id: str, action: SendMessageAction):
         return action_tokens
 
     def get_waiting_test(self, waiting_on: str) -> Optional[str]:
-        assert waiting_on in ["tokens", "time"]
+        assert waiting_on in ["tokens", "time", "percentage_finished"]
         waiting_tests = dict()
         for unique_id, wait_dict in self.wait_list.items():
-            if waiting_on not in wait_dict:
+            is_waiting, test_waiting_on = self.is_waiting(unique_id)
+            if not is_waiting or test_waiting_on != waiting_on:
                 continue
-            waiting_ref = dict(
-                time=datetime.now(),
-                tokens=self.current_token_count,
-            )[waiting_on]
-            if wait_dict[waiting_on] > waiting_ref:
-                waiting_tests[unique_id] = wait_dict
+            waiting_tests[unique_id] = wait_dict
+            #
+            # waiting_ref = dict(
+            #     time=datetime.now(),
+            #     tokens=self.current_token_count,
+            #     percentage_finished=self.percentage_finished
+            # )[waiting_on]
+            # if wait_dict[waiting_on] > waiting_ref:
         if len(waiting_tests) == 0:
             return
         waiting_ids = list(waiting_tests.keys())
         waiting_ids.sort(key=lambda i: waiting_tests[i][waiting_on])
         return waiting_ids[0]
 
-    def is_waiting(self, unique_id: str, remove: bool = False) -> bool:
+    def is_waiting(self, unique_id: str, remove: bool = False) -> Tuple[bool, str]:
         if unique_id not in self.wait_list:
-            return False
+            return False, ""
         wait_dict = self.wait_list[unique_id]
-        if "tokens" in wait_dict and wait_dict["tokens"] > self.current_token_count:
-            return True
-        if "time" in wait_dict and wait_dict["time"] > datetime.now():
-            return True
+        if wait_dict["tokens"] > self.current_token_count:
+            return True, "tokens"
+        if wait_dict["time"] > datetime.now():
+            return True, "time"
+        if wait_dict["percentage_finished"] > self.percentage_finished:
+            return True, "percentage_finished"
         if remove:
             del self.wait_list[unique_id]
-        return False
+        return False, ""
```

Comment:
Actually, looking to the new logic, it doesn't seem like you'd need to change anything for it to be compatible with multi-condition waits.

Reply:
So I put this in, because the logic in `get_waiting_test` seemed to mostly duplicate `is_waiting` with the exception that `is_waiting` didn't specify what you were waiting for. Where `get_waiting_test` is called is in 'pick_next_test_id', the test test will only be resumed if it passes `is_waiting` which checks all of them.

Reply:
The logic is the same, just shorter. `get_waiting_test` will also return a test waiting on a single attribute, regardless of all others. Even in the previous version.  

The code:
```
    for unique_id, wait_dict in self.wait_list.items():
            if waiting_on not in wait_dict:
                continue
            waiting_ref = dict(
                time=datetime.now(),
                tokens=self.current_token_count,
            )[waiting_on]
            if wait_dict[waiting_on] > waiting_ref:
                waiting_tests[unique_id] = wait_dict
```
Is a filter that checks waiting tests, that adds it to a list if the test is still waiting for that attribute. Which is something that `is_waiting` already does. So I just modified `is_waiting` to not only say if a test is waiting, but also one of the attributes that the test is waiting on.

Reply:
Ah right... so before the priorities were taken care of outside, in `pick_next_test_id`, and now it's part of the `get_waiting_test` function. I see the point now. I'll have to think more about the consequences, but yeah, I reckon it might work just fine.